trin-core/src/portalnet/overlay_service.rs

```diff
@@ -2054,26 +2053,21 @@ const TALK_REQ_PACKET_OVERHEAD: usize = 16 + // IV
     9 + // Request ID, max 8 bytes + 1 for RLP encoding
     3 + // RLP Encoding of inner response
     16; // RLP HMAC
-        // ENR SSZ overhead empirically observed to be double.
-        // Todo: determine why this is. It seems too high.
-const MAX_NODES_SIZE: usize = MAX_DISCV5_PACKET_SIZE - TALK_REQ_PACKET_OVERHEAD;
-const NUM_MAX_SIZE_ENRS_IN_MAX_SIZE_PACKET: usize = MAX_NODES_SIZE / MAX_ENR_SIZE;
+const NODES_PACKET_OVERHEAD: usize = 1 + // Selector byte
+    1; // `total` field
+const CONTENT_PACKET_OVERHEAD: usize = 1 + // Selector byte
+    4; // Union type index
+const MAX_NODES_SIZE: usize =
+    MAX_DISCV5_PACKET_SIZE - TALK_REQ_PACKET_OVERHEAD - NODES_PACKET_OVERHEAD;
+const MAX_CONTENT_NODES_SIZE: usize =
+    MAX_DISCV5_PACKET_SIZE - TALK_REQ_PACKET_OVERHEAD - CONTENT_PACKET_OVERHEAD;
 
 /// Limits a to a maximum packet size, including the discv5 header overhead.
-fn limit_enr_list_to_max_bytes(enrs: Vec<SszEnr>, max_size: usize) -> Vec<SszEnr> {
-    // If all ENRs would fit at max size, don't check individual sizes.
-    if enrs.len() < NUM_MAX_SIZE_ENRS_IN_MAX_SIZE_PACKET {
-        return enrs;
-    }
-
-    let mut total_size: usize = 0;
-    enrs.into_iter()
-        .take_while(|enr| {
-            let enr_size = enr.ssz_bytes_len();
-            total_size = total_size + enr_size;
-            total_size < max_size
-        })
-        .collect()
+fn limit_enr_list_to_max_bytes(mut enrs: Vec<SszEnr>, max_size: usize) -> Vec<SszEnr> {
+    while enrs.ssz_bytes_len() > max_size {
+        enrs.pop();
+    }
+    enrs
```

Comment:
Hmm, now this function signature seems a bit unclear. We take a mutable `Vec` and modify it, but we also return the modified `Vec`. I think we should do one or the other but not both. It seems more efficient to modify the existing `Vec` as opposed to allocating a new one, but it could be more clear to pass in a non-mutable `Vec` and return the shrunken result. I don't have a preference, but the function name should reflect that choice.

Reply:
@jacobkaufmann I don't think it's unclear because without returning the modified `Vec` its owernship would never be transferred back to the calling function and we'd get compile errors trying to use it. 

That said, for efficiency purposes I think we should pass by reference which does allow us to just pass in / modify the `Vec` without having to return it. But I'm not sure if "borrow instead of move" was your point. Let me know if I'm misunderstanding.

> but it could be more clear to pass in a non-mutable Vec and return the shrunken result

This strikes me as less clear because as a reader I'd just be wondering why an inefficient approach was taken instead of an easy, more efficient approach

> but the function name should reflect that choice.

Not sure what you mean, do you have a suggestion? `limit_enr_list_to_max_bytes_without_allocating_a_cloned_list_of_enrs`?

Reply:
Sorry, that was an unclear comment. And after a second look, the original implementation with the move of the ENR list makes sense. A more clear comment would have been on move versus mutable borrow. That said, I expect the reader's primary concern to be the clarity of the function's behavior, not its efficiency.

To me, a more descriptive name would be something like `pop_while_ssz_bytes_len_gt`.